pkg/kafka/encoding.go

```diff
@@ -0,0 +1,176 @@
+// Package kafka provides encoding and decoding functionality for Loki's Kafka integration.
+package kafka
+
+import (
+	"errors"
+	"fmt"
+	math_bits "math/bits"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+)
+
+// Encoder is responsible for encoding logproto.Stream data into Kafka records.
+// It handles splitting of large streams into multiple records.
+type Encoder struct {
+	batch logproto.Stream
+}
+
+// NewEncoder creates a new Encoder.
+func NewEncoder() *Encoder {
+	return &Encoder{}
+}
+
+// Encode converts a logproto.Stream into one or more Kafka records.
+// It handles splitting large streams into multiple records if necessary.
+//
+// The encoding process works as follows:
+// 1. If the stream size is smaller than maxSize, it's encoded into a single record.
+// 2. For larger streams, it splits the entries into multiple batches, each under maxSize.
```

Comment:
Is it possible for a single entry to exceed `maxSize`? What happens in that case?